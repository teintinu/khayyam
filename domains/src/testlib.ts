import { createJobManager, JobCallbackArgs, JobManager, Job, Node } from './job'
import { asap, defer } from 'pjobs'
import { Bundler } from './bundler'
import { ByPackage, createWorkspace, Package, WalkedJobs } from './workspace'
import { createProgress } from './progress'
import { System } from './sys'

export const cwd = process.cwd()
const fakeProcess = './scripts/fake-process.js'
const shell = true

export type TreeOfNode = [
  job: string,
  dependencies: TreeOfNode[],
]

export type Logger=ReturnType<typeof createFakeLog>

export function createFakeLog (verbose = false) {
  const logged: string[] = []
  const aOk = defer<void>()
  const bOk = defer<void>()
  const cOk = defer<void>()
  const handleOutput = jest.fn<void, JobCallbackArgs>()
    .mockImplementation(async (job, chunk, error) => {
      const lines = chunk.trim().split('\n')
      lines.forEach((s) => {
        if (s === 'a' || s === 'INPUT: a') setTimeout(aOk.resolve, 100)
        if (s === 'b' || s === 'INPUT: b') setTimeout(bOk.resolve, 100)
        if (s === 'c' || s === 'INPUT: c') setTimeout(cOk.resolve, 100)
        job.progress.update({
          state: 'working',
          message: (error ? ' (error): ' : '') + s
        })
      })
    })
  const jobManager = createJobManager()
  const fsListeners = new Set<()=>void>()
  let tmChanged: any
  const fakeSys: System & { simulateChange():void} = {
    workspaceChanged: fsChanged,
    listenForWorkspaceChanges (callback) {
      fsListeners.add(callback)
      return () => {
        fsListeners.delete(callback)
      }
    },
    getRepository: fsGetRepository,
    simulateChange: fsSimulateChange,
    loadWorkspace,
    notify: fsNotify
  }
  const fakeLogger = {
    jobManager,
    handleOutput,
    aOk,
    bOk,
    cOk,
    fakeSys,
    verbose,
    log (...args: string[]) {
      logged.push(args.join(' '))
      if (verbose) { console.info(args.join(' ')) }
    },
    logged,
    tree () {
      const done: {[id: number]:boolean} = {}
      const tree = jobManager.getTree()
      return tree.map(treeOfNode)
      function treeOfNode (node: Node): TreeOfNode {
        const id = node.job.id
        if (done[id]) {
          return [
            node.job.title + '(rec)',
            []
          ]
        }
        done[id] = true
        const n: TreeOfNode = [
          node.job.title + '=>[' +
          node.dependents.map(d => d.job.title).join() + ']',
          node.dependencies.map(treeOfNode)
        ]
        done[id] = false
        return n
      }
    }
  }
  return fakeLogger
  function fsChanged (): void {
    clearTimeout(tmChanged)
    tmChanged = setTimeout(() => {
      fsListeners.forEach((fn) => asap(fn))
    }, 1000)
  }
  function fsGetRepository (folder: string): string|undefined {
    return 'repo:' + folder
  }
  function fsSimulateChange ():void {
    fsChanged()
  }
  function loadWorkspace (folder:string) {
    const [sBundler, sPkg, sDeps, sLayer, sInvalid] = folder.split('/')
    let bundler:Bundler
    if (sBundler === 'npm') bundler = createFakeBundlerNPM(fakeLogger)
    else throw new Error('fake-fs-invalid-bundle')
    return createFakeWorkspace({
      logger: fakeLogger,
      bundler,
      packages: sPkg as FakePackages,
      deps: sDeps === 'deps',
      layers: sLayer === 'layers',
      invalid: sInvalid === 'invalid'
    })
  }
  function fsNotify (msg: string, pkgName: string) {
    fakeLogger.log('fsNotify:' + msg + ' on package: ' + pkgName)
  }
}

export function createFakeProgress (logger: Logger, title: string, manual: boolean) {
  const pg = createProgress(manual)
  pg.on((status) => {
    logger.log(title + ' ' + status.state + ' ' + status.message)
  })
  return pg
}

export function createFakeJob (logger: Logger, delay: number, title: string, args: string[], manual: boolean = false) {
  logger.log('creating job [' + title + ']')
  const job = logger.jobManager.createJob({
    title,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...args],
    shell,
    queue: 'sequential',
    manual
  })
  job.progress.on((status) => {
    logger.log('pg ' + title + ': ' + status.state + ' ' + status.message)
  })
  job.listen(logger.handleOutput)
  return job
}

export function create2FakeJob (logger: Logger, delay: number, mainTitle: string, mainArgs: string[], depTitle: string, depArgs: string[]) {
  logger.log('creating main job [' + mainTitle + ']')
  const main = logger.jobManager.createJob({
    title: mainTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...mainArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  logger.log('creating dep job [' + depTitle + ']')
  const dep = logger.jobManager.createJob({
    title: depTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...depArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  main.depends(dep)
  main.progress.on((status) => {
    logger.log('pg-main ' + status.state + ' ' + status.message)
  })
  dep.progress.on((status) => {
    logger.log('pg-dep ' + status.state + ' ' + status.message)
  })
  main.listen(logger.handleOutput)
  dep.listen(logger.handleOutput)
  return {
    main, dep
  }
}

function createFakeBundlerNPM (logger:Logger) {
  const fakeBundler: Bundler = {
    name: 'fakeBundlerNPM',
    watch (pkg) {
      return ['fakeBundlerNPM']
    },
    build (pkg, jobManager, goal) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-build-' + goal)
    },
    test (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-test')
    },
    serve (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-serve')
    },
    lint (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-lint')
    },
    publish (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-publish')
    }
  }
  return fakeBundler
  function createFakeBundlerCommand (pkg: Package, cmdJobManager: JobManager, cmd: string) {
    expect(cmdJobManager).toBe(logger.jobManager)
    return createFakeJob(
      logger,
      100,
      pkg.name + '(' + cmd + ')',
      [
        cmd
      ]
    )
  }
}

type FakePackages='x'|'xy'|'abcde'

function createFakeWorkspace (
  { logger, bundler, packages, deps, layers, invalid }:
  { logger: Logger, bundler: Bundler, packages:FakePackages, deps: boolean; layers: boolean; invalid: boolean }) {
  const pkgs:{
      [n:string]:Package
    } = {}
  for (const n of packages) {
    const pkg: Readonly<Package > = {
      name: n,
      layer: layers ? n : '',
      dependencies: [],
      bundlers: [bundler.name],
      folder: n
    }
    pkgs[n] = pkg
  }
  if (deps) {
    if (packages === 'xy') {
      pkgs.x = {
        ...pkgs.x,
        dependencies: invalid ? ['z'] : ['y']
      }
    }
    if (packages === 'abcde') {
      pkgs.a = {
        ...pkgs.a,
        dependencies: invalid ? ['b', 'c', 'z'] : ['b', 'c']
      }
      pkgs.b = {
        ...pkgs.b,
        dependencies: ['c', 'd']
      }
      pkgs.c = {
        ...pkgs.c,
        dependencies: ['d']
      }
      pkgs.d = {
        ...pkgs.d,
        dependencies: ['e']
      }
    }
  }
  const ws = createWorkspace({
    sys: logger.fakeSys,
    layers: [],
    packages: Object.keys(pkgs).map((n) => pkgs[n]),
    bundlers: [bundler]
  })
  return ws
}

export function jobDebug (job: Job) {
  return job.title + '(' +
    [...job.dependencies]
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(dep => dep.title) +
    ')'
}

export function walkedDebug (walkJobs: WalkedJobs) {
  const ret: ByPackage<unknown> = {}
  const pkgs = Object.keys(walkJobs.jobs).sort()
  pkgs.forEach((pkg) => {
    ret[pkg] = ([...walkJobs.jobs[pkg]] || [])
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(jobDebug)
  })
  return ret
}
