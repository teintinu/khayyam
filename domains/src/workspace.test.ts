import { createFakeJob, createFakeLog, walkedDebug } from './testlib'

describe('workspace', () => {
  describe('just one package', () => {
    it('create workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/x')
      expect(ws.packages.map((p) => p.name).sort().join()).toBe('x')
      expect(ws.bundlers.map((b) => b.name).sort().join()).toBe('fakeBundlerNPM')
      expect(logger.logged?.length).toBe(0)
      expect({
        layers: ws.layers,
        pkgs: ws.packages
      }).toMatchSnapshot()
    })
    it('findPackage workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/x')
      expect(ws.findPackage('x')?.name).toBe('x')
      expect(ws.findPackage('a')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    it('findBundler workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/x')
      expect(ws.findBundler('fakeBundlerNPM')?.name).toBe('fakeBundlerNPM')
      expect(ws.findBundler('invalid')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    describe('walk workspace', () => {
      it('all', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/x')
        const pkgs: string[] = []
        const walked = ws.walk('all', (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walk',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('x')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter x', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/x')
        const pkgs: string[] = []
        const walked = ws.walk({ x: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walk',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('x')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter none', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/x')
        const pkgs: string[] = []
        const walked = ws.walk({}, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walk',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
    })
  })
  describe('just x,y packages', () => {
    it('create workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
      expect(ws.packages.map((p) => p.name).sort().join()).toBe('x,y')
      expect(ws.bundlers.map((b) => b.name).sort().join()).toBe('fakeBundlerNPM')
      expect(logger.logged?.length).toBe(0)
      expect({
        layers: ws.layers,
        pkgs: ws.packages
      }).toMatchSnapshot()
    })
    it('findPackage workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
      const x = ws.findPackage('x')
      expect(x?.name).toBe('x')
      expect(x?.dependencies.join()).toBe('y')
      expect(ws.findPackage('a')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    it('findBundler workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
      expect(ws.findBundler('fakeBundlerNPM')?.name).toBe('fakeBundlerNPM')
      expect(ws.findBundler('invalid')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    describe('walk workspace', () => {
      it('all', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
        const pkgs: string[] = []
        const walked = ws.walk('all', (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkXY',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('y,x')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter x', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ x: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkXY',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('x')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter y', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ y: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkXY',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('y,x')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter none', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/xy/deps')
        const pkgs: string[] = []
        const walked = ws.walk({}, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkXY',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
    })
  })
  describe('just a,b,c,d,e packages', () => {
    it('create workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
      expect(ws.packages.map((p) => p.name).sort().join()).toBe('a,b,c,d,e')
      expect(ws.bundlers.map((b) => b.name).sort().join()).toBe('fakeBundlerNPM')
      expect(logger.logged?.length).toBe(0)
      expect({
        layers: ws.layers,
        pkgs: ws.packages
      }).toMatchSnapshot()
    })
    it('findPackage workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
      const a = ws.findPackage('a')
      expect(a?.name).toBe('a')
      expect(a?.dependencies.join()).toBe('b,c')
      expect(ws.findPackage('x')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    it('findBundler workspace', async () => {
      const logger = createFakeLog()
      const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
      expect(ws.findBundler('fakeBundlerNPM')?.name).toBe('fakeBundlerNPM')
      expect(ws.findBundler('invalid')).toBeUndefined()
      expect(logger.logged?.length).toBe(0)
    })
    describe('walk workspace', () => {
      it('all', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk('all', (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('e,d,c,b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter a', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ a: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter b', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ b: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter c', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ c: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('c,b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter d', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ d: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('d,c,b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter c,d', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ c: true, d: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('d,c,b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter e', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({ e: true }, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('e,d,c,b,a')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
      it('filter none', () => {
        const logger = createFakeLog()
        const ws = logger.fakeSys.loadWorkspace('npm/abcde/deps')
        const pkgs: string[] = []
        const walked = ws.walk({}, (pkg) => {
          pkgs.push(pkg.name)
          return createFakeJob(
            logger,
            10,
            pkg.name + ' walkABCDE',
            ['walk']
          )
        })
        expect(pkgs.join()).toBe('')
        expect(logger.logged?.length).toBe(0)
        expect(walkedDebug(walked)).toMatchSnapshot('walkedDebug')
      })
    })
  })
})
