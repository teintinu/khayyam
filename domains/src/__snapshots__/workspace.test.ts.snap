// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workspace just a,b,c,d,e packages create workspace: layers/pkg 1`] = `
Object {
  "layers": Array [],
  "pkgs": Array [
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [
        "b",
        "c",
      ],
      "folder": "a",
      "layer": "",
      "name": "a",
    },
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [
        "c",
        "d",
      ],
      "folder": "b",
      "layer": "",
      "name": "b",
    },
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [
        "d",
      ],
      "folder": "c",
      "layer": "",
      "name": "c",
    },
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [
        "e",
      ],
      "folder": "d",
      "layer": "",
      "name": "d",
    },
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [],
      "folder": "e",
      "layer": "",
      "name": "e",
    },
  ],
}
`;

exports[`workspace just a,b,c,d,e packages create workspace: logged 1`] = `Array []`;

exports[`workspace just a,b,c,d,e packages findBundler workspace: logged 1`] = `Array []`;

exports[`workspace just a,b,c,d,e packages findPackage workspace: logged 1`] = `Array []`;

exports[`workspace just a,b,c,d,e packages walk workspace all: logged 1`] = `
Array [
  "creating job [e walkABCDE]",
  "creating job [d walkABCDE]",
  "creating job [c walkABCDE]",
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace all: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE,c walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE(c walkABCDE,d walkABCDE)",
  ],
  "c": Array [
    "c walkABCDE(d walkABCDE)",
  ],
  "d": Array [
    "d walkABCDE(e walkABCDE)",
  ],
  "e": Array [
    "e walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter a: logged 1`] = `
Array [
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter a: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter b: logged 1`] = `
Array [
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter b: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter c,d: logged 1`] = `
Array [
  "creating job [d walkABCDE]",
  "creating job [c walkABCDE]",
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter c,d: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE,c walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE(c walkABCDE,d walkABCDE)",
  ],
  "c": Array [
    "c walkABCDE(d walkABCDE)",
  ],
  "d": Array [
    "d walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter c: logged 1`] = `
Array [
  "creating job [c walkABCDE]",
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter c: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE,c walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE(c walkABCDE)",
  ],
  "c": Array [
    "c walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter d: logged 1`] = `
Array [
  "creating job [d walkABCDE]",
  "creating job [c walkABCDE]",
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter d: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE,c walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE(c walkABCDE,d walkABCDE)",
  ],
  "c": Array [
    "c walkABCDE(d walkABCDE)",
  ],
  "d": Array [
    "d walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter e: logged 1`] = `
Array [
  "creating job [e walkABCDE]",
  "creating job [d walkABCDE]",
  "creating job [c walkABCDE]",
  "creating job [b walkABCDE]",
  "creating job [a walkABCDE]",
]
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter e: walkedDebug 1`] = `
Object {
  "a": Array [
    "a walkABCDE(b walkABCDE,c walkABCDE)",
  ],
  "b": Array [
    "b walkABCDE(c walkABCDE,d walkABCDE)",
  ],
  "c": Array [
    "c walkABCDE(d walkABCDE)",
  ],
  "d": Array [
    "d walkABCDE(e walkABCDE)",
  ],
  "e": Array [
    "e walkABCDE()",
  ],
}
`;

exports[`workspace just a,b,c,d,e packages walk workspace filter none: logged 1`] = `Array []`;

exports[`workspace just a,b,c,d,e packages walk workspace filter none: walkedDebug 1`] = `Object {}`;

exports[`workspace just one package create: layers/pkg 1`] = `
Object {
  "layers": Array [],
  "pkgs": Array [
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [],
      "folder": "x",
      "layer": "",
      "name": "x",
    },
  ],
}
`;

exports[`workspace just one package create: logged 1`] = `Array []`;

exports[`workspace just one package findBundler: logged 1`] = `Array []`;

exports[`workspace just one package findPackage: logged 1`] = `Array []`;

exports[`workspace just one package walk one job per package all: logged 1`] = `
Array [
  "creating job [x walk]",
]
`;

exports[`workspace just one package walk one job per package all: walkedDebug 1`] = `
Object {
  "x": Array [
    "x walk()",
  ],
}
`;

exports[`workspace just one package walk one job per package filter none: logged 1`] = `Array []`;

exports[`workspace just one package walk one job per package filter none: walkedDebug 1`] = `Object {}`;

exports[`workspace just one package walk one job per package filter x: logged 1`] = `
Array [
  "creating job [x walk]",
]
`;

exports[`workspace just one package walk one job per package filter x: walkedDebug 1`] = `
Object {
  "x": Array [
    "x walk()",
  ],
}
`;

exports[`workspace just one package walk two jobs per package all: logged 1`] = `
Array [
  "creating job [x walk1]",
  "creating job [x walk2]",
]
`;

exports[`workspace just one package walk two jobs per package all: walked1Debug 1`] = `
Object {
  "x": Array [
    "x walk1()",
  ],
}
`;

exports[`workspace just one package walk two jobs per package all: walked2Debug 1`] = `
Object {
  "x": Array [
    "x walk2()",
  ],
}
`;

exports[`workspace just one package walk two jobs per package filter none: logged 1`] = `Array []`;

exports[`workspace just one package walk two jobs per package filter none: walked1Debug 1`] = `Object {}`;

exports[`workspace just one package walk two jobs per package filter none: walked2Debug 1`] = `Object {}`;

exports[`workspace just one package walk two jobs per package filter x: logged 1`] = `
Array [
  "creating job [x walk1]",
  "creating job [x walk2]",
]
`;

exports[`workspace just one package walk two jobs per package filter x: walked1Debug 1`] = `
Object {
  "x": Array [
    "x walk1()",
  ],
}
`;

exports[`workspace just one package walk two jobs per package filter x: walked2Debug 1`] = `
Object {
  "x": Array [
    "x walk2()",
  ],
}
`;

exports[`workspace just x,y packages create workspace: layers/pkg 1`] = `
Object {
  "layers": Array [],
  "pkgs": Array [
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [
        "y",
      ],
      "folder": "x",
      "layer": "",
      "name": "x",
    },
    Object {
      "bundlers": Array [
        "fakeBundlerNPM",
      ],
      "dependencies": Array [],
      "folder": "y",
      "layer": "",
      "name": "y",
    },
  ],
}
`;

exports[`workspace just x,y packages create workspace: logged 1`] = `Array []`;

exports[`workspace just x,y packages findBundler workspace: logged 1`] = `Array []`;

exports[`workspace just x,y packages findPackage workspace: logged 1`] = `Array []`;

exports[`workspace just x,y packages walk workspace all: logged 1`] = `
Array [
  "creating job [y walkXY]",
  "creating job [x walkXY]",
]
`;

exports[`workspace just x,y packages walk workspace all: walkedDebug 1`] = `
Object {
  "x": Array [
    "x walkXY(y walkXY)",
  ],
  "y": Array [
    "y walkXY()",
  ],
}
`;

exports[`workspace just x,y packages walk workspace filter none: logged 1`] = `Array []`;

exports[`workspace just x,y packages walk workspace filter none: walkedDebug 1`] = `Object {}`;

exports[`workspace just x,y packages walk workspace filter x: logged 1`] = `
Array [
  "creating job [x walkXY]",
]
`;

exports[`workspace just x,y packages walk workspace filter x: walkedDebug 1`] = `
Object {
  "x": Array [
    "x walkXY()",
  ],
}
`;

exports[`workspace just x,y packages walk workspace filter y: logged 1`] = `
Array [
  "creating job [y walkXY]",
  "creating job [x walkXY]",
]
`;

exports[`workspace just x,y packages walk workspace filter y: walkedDebug 1`] = `
Object {
  "x": Array [
    "x walkXY(y walkXY)",
  ],
  "y": Array [
    "y walkXY()",
  ],
}
`;
